logstash-core/src/main/java/org/logstash/ackedqueue/io/FileCheckpointIO.java

```diff
@@ -115,14 +116,18 @@ private Checkpoint read(BufferedChecksumStreamInput crcsi) throws IOException {
         return new Checkpoint(pageNum, firstUnackedPageNum, firstUnackedSeqNum, minSeqNum, elementCount);
     }
 
-    private void write(Checkpoint checkpoint, byte[] buf) throws IOException {
-        BufferedChecksumStreamOutput output = new BufferedChecksumStreamOutput(new ByteArrayStreamOutput(buf));
-        output.writeShort((short)Checkpoint.VERSION);
-        output.writeInt(checkpoint.getPageNum());
-        output.writeInt(checkpoint.getFirstUnackedPageNum());
-        output.writeLong(checkpoint.getFirstUnackedSeqNum());
-        output.writeLong(checkpoint.getMinSeqNum());
-        output.writeInt(checkpoint.getElementCount());
-        output.writeInt((int)output.getChecksum());
+    private void write(Checkpoint checkpoint, ByteBuffer buf) {
+        crc32.reset();
+        buf.clear();
+        buf.putShort((short)Checkpoint.VERSION);
+        buf.putInt(checkpoint.getPageNum());
+        buf.putInt(checkpoint.getFirstUnackedPageNum());
+        buf.putLong(checkpoint.getFirstUnackedSeqNum());
+        buf.putLong(checkpoint.getMinSeqNum());
+        buf.putInt(checkpoint.getElementCount());
+        buf.flip();
+        crc32.update(buf);
+        buf.position(BUFFER_SIZE - Integer.BYTES).limit(BUFFER_SIZE);
```

Comment:
the bytebuffer api is such a strange thing.

I'd kind of prefer using `buf.mark()` (before you call `flip`) then `buf.reset()` instead of `buf.position(...)` to avoid any math problems, but I leave it up to you to decide. (keep the `.limit` call)